src/Controls/src/Core/NavigationPage/NavigationPage.cs

```diff
@@ -575,7 +575,23 @@ void OnWindowChanged(object sender, EventArgs e)
 				else
 				{
 					// Is the root the window or is this part of a modal stack
-					var rootPage = this.FindParentWith(x => (x is IWindow te || Window.Navigation.ModalStack.Contains(x)), true);
+					Element rootPage;
+
+					var parentPages = this.GetParentPages();
+					parentPages.Insert(0, this);
+
+					// Is my top parent page the root page on the window?
+					// If so then we set the toolbar on the window
+					if (Window.Page == parentPages[parentPages.Count - 1])
+					{
+						rootPage = Window;
+					}
+					else
+					{
+						// This means the page is a modal page so we set the toolbar on the top level page
+						// of the modal
+						rootPage = parentPages[parentPages.Count - 1];
```

Comment:
is this the best way to make sure it's modal? also we could grab this topParentPage for the check and assign 1 time earlier .

Reply:
> also we could grab this topParentPage for the check and assign 1 time earlier .

fixed

> is this the best way to make sure it's modal

Maybe? Before this PR it was checking `ModalStack` to see if its parent page was inside `ModalStack` but the order everything was firing in would cause this code to trigger before being added to the `ModalStack`. This would make the code thing it was on the root window which would then break everything, so I'm trying to decouple this code a bit from external dependencies. 

Now it just relies on its own page hierarchy to locate its top parent and it's not dependent on external structures being setup before reaching this code.